Document how to claim a cluster for a job (#123)
openshift/ci-docs: content/en/docs/how-tos/cluster-claim.md

---
title: "Creating a Cluster Pool"
description: How to create a cluster pool from which a job can claim a cluster.
---

> DPTP integrates Hive into the OpenShift test platform to manage the cluster pools and clusters in them. There is a defined pool [ci-cluster-pool](https://github.com/openshift/release/tree/master/clusters/app.ci/ci-cluster-pool) from which can be used to claim an OpenShift cluster for a test. In this page, we document how to create a new cluster pool with the cloud account provided by a CI user.

Reviewer: IMO we should really be more user oriented here: what problem does this doc solve. The first paragraph should re-assure the user they came to the right place. The user does not care at all about DPTP or HIve. The user is very likely an "owner" or a lead of some group or team, they have resources at their disposal (cloud accounts), and they want to set things so that their jobs can use these accounts. The paragraph should be about that.

## Prerequisites

> *  Choose a [Hive](https://github.com/openshift/hive#supported-cloud-providers) supported cloud for the pool.

Reviewer: ```suggestion
* Choose a [Hive](https://github.com/openshift/hive#supported-cloud-providers) supported cloud for the pool.
```

Reviewer: And probably more something like "Vefirify that the cloud platform you want to use is supported by Hive".

* [Configure the cloud account](https://docs.openshift.com/container-platform/4.7/installing/installing-preparing.html).

## Create a cluster pool

> Create a folder in `release/clusters/app.ci/` with the manifests for the pool. See [the examples](https://github.com/openshift/release/tree/master/clusters/app.ci/ci-cluster-pool) for the existing `ci-cluster-pool`.

Reviewer: Hrm, maybe we want to group them? IMO, we should optimize for a case where every pool-owning group has multiple pools, and we have multiple (or even many) groups that manage their pools. So IMO we want this:

```
clusters/
  app.ci/
    pools/
      groupA/
        pool1.yaml
        pool2.yaml
        ...
      groupB/
        pool1.yaml
        pool2.yaml
```

and not 

```
clusters/
  app.ci/
    groupA/
      pool1.yaml
      pool2.yaml
      ...
    groupB/
      pool1.yaml
      pool2.yaml
```

Reviewer: Also `release/` is not a path in `openshift/release`.

Author: https://github.com/openshift/release/pull/17950

```yaml
apiVersion: hive.openshift.io/v1
kind: ClusterPool
metadata:
  name: ci-ocp-4-7-amd64-aws-us-east-1
  namespace: cvp-cluster-pool
  labels: # product, version, architecture, cloud, owner are used to filter out a pool when a job claims a cluster
    product: ocp
    version: "4.7"
    architecture: amd64
    cloud: aws
    owner: cvp
    region: us-east-1
spec:
  baseDomain: hive.example.org # the base domain to install the cluster
  imageSetRef:
    name: ocp-4.7.0-amd64 # the name of the imageSet which determines the image to install the cluster
  installConfigSecretTemplateRef:
    name: install-config-aws-us-east-1 # the name of the secret with an installation config for the installer
  skipMachinePools: true
  platform:
    aws:
      credentialsSecretRef:
        name: hive-aws-credentials # the name of the secret with the credentials of the cloud account
      region: us-east-1
  pullSecretRef:

>     name: pull-secret # the name of the pull secret used by installer

Reviewer: Is this something that a user needs to set? AFAIK we need everyone to use our pull-secrets (so they can pull ci-op built images,etc), no?

Author: No.
We only install GA version of OpenShift via pools. No CI images are needed.

The pull secret can be obtained as described https://docs.openshift.com/container-platform/4.7/installing/installing_aws/installing-aws-default.html

Reviewer: > We only install GA version of OpenShift via pools. No CI images are needed. 

That's not true. If a user wants to test their operator, they will build it in CI and the test cluster must be able to install it.

Author: Interesting. In principle, this isnt the concern of the pool.
This is day 2 operation (after installation), I guess.
The test has admin access to the cluster. They can modify the global pull secret if needed.

For example, my test cases (I know very unlikely) are
Case1: my operator pod does not run without linking the pull secret to the service account running the pod.

Case2: my operator pod runs after linking the pull secret to the service account running the pod.

I just realize that this become tricky.
Where are the images built in CI for an operator used?
- image of a pod
- input of a build
- something else

How are they referred?
- build01.registry.ci.openshft.org/ci-op-xxxx/component:tag
- image-registry.openshift-image-registry.svc:5000/ci-op-xxxx/component:tag

In the case of latter, it might be hard to inject it in advance because it is not clear from which cluster the token should be taken.

Author: I saw many templates make the images in ci-op-xxxx public.
We could do the same in for oo tests.

Author: How come this works?
https://github.com/openshift/release/pull/17798/files
Does it install any image built in CI?

Reviewer: Huh, right. IIRC there are steps that expose the images, indeed. Perfect, let's keep it this way. At worst we iterate.

  size: 1
  maxSize: 10

> ```

Reviewer: Comment the `size` and `maxSize` fields too?

Author: Updated.

> The referred manifests in the pool are in the same folder. The meaning of other fields in the pool specification is available at [Hive's documentation](https://pkg.go.dev/github.com/openshift/hive/apis@master/hive/v1#ClusterPool).

Reviewer: We should IMO at least briefly discuss the sizing and resource consumption.

Author: Updated.

---
title: "Creating a Cluster Pool"
description: How to create a cluster pool from which a job can claim a cluster.
---

> If a CI user wants to use their own cloud account to provision clusters, the steps of creating such a cluster pool can be found in this page.

Reviewer: ```suggestion
This document describes how to set up cluster pools backed by custom cloud platform accounts owned by users or teams that want their jobs to use clusters provisioned in these custom accounts.
```

Author: Updated.

## Prerequisites

> * Vefirify that the cloud platform you want to use is supported by [Hive](https://github.com/openshift/hive#supported-cloud-providers) which is the backbone of cluster pools in the test platform.

Reviewer: ```suggestion
* Verify that the cloud platform you want to use is supported by [Hive](https://github.com/openshift/hive#supported-cloud-providers) (the cluster pool mechanism back-end)
```

Author: Updated.

* [Configure the cloud account](https://docs.openshift.com/container-platform/4.7/installing/installing-preparing.html): Hive does not require more configuration of the cloud account than installing an OpenShift cluster. Make sure that the account has quotas for the pools to contain the desired number of clusters.

## Create a cluster pool

> Create a folder in `clusters/app.ci/pools` for the manifests of the pool. See [openshift-ci](https://github.com/openshift/release/tree/master/clusters/app.ci/pools/openshift-ci) as an example for the existing `ci-cluster-pool`.

Reviewer: ```suggestion
In the openshift/release repository, create a folder in the [`clusters/app.ci/pools`](https://github.com/openshift/release/tree/master/clusters/app.ci/pools) directory that will contain manifests of all your pools (see [openshift-ci](https://github.com/openshift/release/tree/master/clusters/app.ci/pools/openshift-ci) as an example). Then place the manifests for the new `ClusterPool` there.
```

Author: Updated.

```yaml
apiVersion: hive.openshift.io/v1
kind: ClusterPool
metadata:
  name: ci-ocp-4-7-amd64-aws-us-east-1
  namespace: cvp-cluster-pool
  labels: # product, version, architecture, cloud, owner are used to filter out a pool when a job claims a cluster
    product: ocp
    version: "4.7"
    architecture: amd64
    cloud: aws
    owner: cvp
    region: us-east-1
spec:
  baseDomain: hive.example.org # the base domain to install the cluster
  imageSetRef:
    name: ocp-4.7.0-amd64 # the name of the imageSet which determines the image to install the cluster
  installConfigSecretTemplateRef:
    name: install-config-aws-us-east-1 # the name of the secret with an installation config for the installer
  skipMachinePools: true
  platform:
    aws:
      credentialsSecretRef:
        name: hive-aws-credentials # the name of the secret with the credentials of the cloud account
      region: us-east-1
  pullSecretRef:
    name: pull-secret # the name of the pull secret used by installer
  size: 1 # the number of clusters that Hive should keep provisioned and waiting for use.
  maxSize: 10 # the maximum number of clusters that have not been destroyed.

> ```

Reviewer: This only describes one of the necessary resources, but the existing folder contains more - a secret used as an install config, a clusterimageset, a namespace? The instructions should discuss them.

Author: Updated.
